File: go/vcard.go
package nexcore

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

// VCardNamespace implements the v1 虚拟信用卡 endpoints.
//
// 对应 NexCore 虚拟信用卡模块(对照 internal/handler/virtual_card_api.go):
//
// 只读 / 普通写(X-API-Key + X-Secret-Key 双 header):
//
//	GET  /api/v1/vcard/info                       GetInfo              平台公开信息
//	GET  /api/v1/vcard/bins                        ListBins             可开卡 BIN 列表
//	GET  /api/v1/vcard/cards                       ListCards            我的卡片列表
//	GET  /api/v1/vcard/cards/{id}/transactions     GetCardTransactions  卡片消费流水
//	GET  /api/v1/vcard/orders                      ListOrders           订单列表
//	GET  /api/v1/vcard/orders/{id}                 GetOrder             单笔订单
//	PUT  /api/v1/vcard/cards/{id}/remark           UpdateCardRemark     修改卡备注
//
// 资金 / 敏感(HMAC-SHA256 签名 + X-Key-ID/X-Timestamp/X-Nonce/X-Signature):
//
//	GET  /api/v1/vcard/cards/{id}/details          GetCardDetails       卡敏感信息(卡号等)
//	GET  /api/v1/vcard/cards/{id}/code             GetCardCode          CVV / 安全码
//	POST /api/v1/vcard/cards                        OpenCard             开卡
//	POST /api/v1/vcard/cards/{id}/recharge          RechargeCard         充值
//	POST /api/v1/vcard/cards/{id}/cancel            CancelCard           注销
//
// 鉴权密钥来自 cfg.APIKey / cfg.APISecret(MPK 商户密钥),与 account 命名空间共用.
type VCardNamespace struct {
	c *Client
}

// headers 构造 X-API-Key + X-Secret-Key(双密钥读 / 普通写,复用 energy 模式).
func (n *VCardNamespace) headers() (map[string]string, error) {
	if n.c.cfg.APIKey == "" || n.c.cfg.APISecret == "" {
		return nil, &Error{Message: "APIKey / APISecret not configured", Code: -1}
	}
	return map[string]string{
		"X-API-Key":    n.c.cfg.APIKey,
		"X-Secret-Key": n.c.cfg.APISecret,
	}, nil
}

// signedRequest 统一处理 HMAC 签名头 + BodyRaw 发送.
//
// 严格保证签名串里的 body 与实际发出的字节完全一致:bodyBytes 既参与签名,
// 又通过 requestOpts.BodyRaw 原样发出(像 withdraw 那样),杜绝 http 层二次序列化
// 导致签名不匹配.GET 请求 bodyBytes 传 nil,签名里 body 视为空字符串.
//
// 签名公式(与后端 internal/middleware/api_auth.go verifySignature 字节级一致):
//
//	payload = ts + nonce + method + path + rawQuery + body
//	sig     = hex_lower( HMAC_SHA256(APISecret, payload) )
//
// 这些签名接口都不带 query,rawQuery 固定为 "".
func (n *VCardNamespace) signedRequest(method, path string, bodyBytes []byte) (json.RawMessage, error) {
	if n.c.cfg.APIKey == "" || n.c.cfg.APISecret == "" {
		return nil, &Error{Message: "APIKey / APISecret not configured", Code: -1}
	}

	ts := strconv.FormatInt(time.Now().Unix(), 10)
	nonce, err := newHexNonce()
	if err != nil {
		return nil, &Error{Message: "vcard: gen nonce: " + err.Error(), Code: -1}
	}

	const rawQuery = "" // 签名接口均无 query
	body := ""
	if bodyBytes != nil {
		body = string(bodyBytes)
	}

	payload := ts + nonce + strings.ToUpper(method) + path + rawQuery + body
	mac := hmac.New(sha256.New, []byte(n.c.cfg.APISecret))
	mac.Write([]byte(payload))
	sig := hex.EncodeToString(mac.Sum(nil))

	return n.c.transport.do(method, path, &requestOpts{
		BodyRaw: bodyBytes,
		Headers: map[string]string{
			"X-Key-ID":    n.c.cfg.APIKey,
			"X-Timestamp": ts,
			"X-Nonce":     nonce,
			"X-Signature": sig,
		},
	})
}

// newHexNonce 生成 16 字节(32 hex 字符)随机 nonce,满足后端 nonce 长度要求.
func newHexNonce() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

// ---------- 双密钥读 / 普通写 ----------

// GetInfo returns vcard platform public info.
//
// GET /api/v1/vcard/info
func (n *VCardNamespace) GetInfo() (json.RawMessage, error) {
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("GET", "/api/v1/vcard/info", &requestOpts{Headers: h})
}

// ListBins lists open-card BINs (card platforms / tiers).
//
// GET /api/v1/vcard/bins
func (n *VCardNamespace) ListBins() (json.RawMessage, error) {
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("GET", "/api/v1/vcard/bins", &requestOpts{Headers: h})
}

// ListCards lists the merchant's cards.
//
// GET /api/v1/vcard/cards
func (n *VCardNamespace) ListCards() (json.RawMessage, error) {
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("GET", "/api/v1/vcard/cards", &requestOpts{Headers: h})
}

// GetCardTransactions returns the consumption transactions of a card.
//
// GET /api/v1/vcard/cards/{id}/transactions
func (n *VCardNamespace) GetCardTransactions(cardID string) (json.RawMessage, error) {
	if cardID == "" {
		return nil, &Error{Message: "vcard: cardID is required", Code: -1}
	}
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("GET", fmt.Sprintf("/api/v1/vcard/cards/%s/transactions", cardID), &requestOpts{Headers: h})
}

// ListOrders lists vcard orders.
//
// GET /api/v1/vcard/orders
//
// query 支持 page / page_size / status / order_type.
func (n *VCardNamespace) ListOrders(query map[string]any) (json.RawMessage, error) {
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("GET", "/api/v1/vcard/orders", &requestOpts{Query: query, Headers: h})
}

// GetOrder returns a single vcard order.
//
// GET /api/v1/vcard/orders/{id}
func (n *VCardNamespace) GetOrder(orderID string) (json.RawMessage, error) {
	if orderID == "" {
		return nil, &Error{Message: "vcard: orderID is required", Code: -1}
	}
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("GET", fmt.Sprintf("/api/v1/vcard/orders/%s", orderID), &requestOpts{Headers: h})
}

// UpdateCardRemark updates a card's remark.
//
// PUT /api/v1/vcard/cards/{id}/remark   body {"remark": remark}
//
// 非资金敏感,使用双密钥写(X-API-Key + X-Secret-Key).
func (n *VCardNamespace) UpdateCardRemark(cardID, remark string) (json.RawMessage, error) {
	if cardID == "" {
		return nil, &Error{Message: "vcard: cardID is required", Code: -1}
	}
	h, err := n.headers()
	if err != nil {
		return nil, err
	}
	return n.c.transport.do("PUT", fmt.Sprintf("/api/v1/vcard/cards/%s/remark", cardID), &requestOpts{
		Body:    map[string]any{"remark": remark},
		Headers: h,
	})
}

// ---------- HMAC 签名(资金 / 敏感) ----------

// GetCardDetails returns the card's sensitive details (full PAN, expiry, etc).
//
// GET /api/v1/vcard/cards/{id}/details  — HMAC 签名鉴权.
func (n *VCardNamespace) GetCardDetails(cardID string) (json.RawMessage, error) {
	if cardID == "" {
		return nil, &Error{Message: "vcard: cardID is required", Code: -1}
	}
	return n.signedRequest("GET", fmt.Sprintf("/api/v1/vcard/cards/%s/details", cardID), nil)
}

// GetCardCode returns the card's security code (CVV/CVC).
//
// GET /api/v1/vcard/cards/{id}/code  — HMAC 签名鉴权.
func (n *VCardNamespace) GetCardCode(cardID string) (json.RawMessage, error) {
	if cardID == "" {
		return nil, &Error{Message: "vcard: cardID is required", Code: -1}
	}
	return n.signedRequest("GET", fmt.Sprintf("/api/v1/vcard/cards/%s/code", cardID), nil)
}

// OpenCard opens a new virtual card.
//
// POST /api/v1/vcard/cards  — HMAC 签名鉴权.
//
//	params := map[string]any{
//	    "bin_platform_id": 1,
//	    "amount":          "20.00",
//	}
func (n *VCardNamespace) OpenCard(params map[string]any) (json.RawMessage, error) {
	body, err := json.Marshal(params)
	if err != nil {
		return nil, &Error{Message: "vcard: marshal body: " + err.Error(), Code: -1}
	}
	return n.signedRequest("POST", "/api/v1/vcard/cards", body)
}

// RechargeCard recharges an existing card.
//
// POST /api/v1/vcard/cards/{id}/recharge  — HMAC 签名鉴权.
//
//	params := map[string]any{"amount": "10.00"}
func (n *VCardNamespace) RechargeCard(cardID string, params map[string]any) (json.RawMessage, error) {
	if cardID == "" {
		return nil, &Error{Message: "vcard: cardID is required", Code: -1}
	}
	body, err := json.Marshal(params)
	if err != nil {
		return nil, &Error{Message: "vcard: marshal body: " + err.Error(), Code: -1}
	}
	return n.signedRequest("POST", fmt.Sprintf("/api/v1/vcard/cards/%s/recharge", cardID), body)
}

// CancelCard cancels (closes) a card.
//
// POST /api/v1/vcard/cards/{id}/cancel  — HMAC 签名鉴权,无 body.
func (n *VCardNamespace) CancelCard(cardID string) (json.RawMessage, error) {
	if cardID == "" {
		return nil, &Error{Message: "vcard: cardID is required", Code: -1}
	}
	return n.signedRequest("POST", fmt.Sprintf("/api/v1/vcard/cards/%s/cancel", cardID), nil)
}

// VerifyWebhook 校验平台推送的 webhook 签名(给对接方验证平台主动推送的真实性).
//
// 复刻后端 pkg.GenerateSign 的算法:
//   - 取 params 中所有「非空、且 key != "sign"」的字段
//   - 按 key 升序拼成 "k1=v1&k2=v2&..."
//   - sign = hex_lower( HMAC_SHA256(secret, 拼接串) )
//   - 与 params["sign"] 用 hmac.Equal 做常量时间比较
//
// 防重放:平台推送一般同时带 sign_ts(签名时刻,unix 秒,±300s 窗口)与 nonce(随机串).
// 本函数只校验签名;时间窗(now-sign_ts 在 ±300s 内)与 nonce 去重需由调用方自行实现,
// 二者配合才能完整防重放.
//
// 用法:
//
//	params := map[string]string{
//	    "event":   "card.transaction",
//	    "card_id": "123",
//	    "sign_ts": "1718000000",
//	    "nonce":   "abc123...",
//	    "sign":    req.Header...或 body 里的 sign 字段,
//	}
//	if !nexcore.VerifyWebhook(params, webhookSecret) {
//	    // 验签失败,拒绝处理
//	}
func VerifyWebhook(params map[string]string, secret string) bool {
	got := params["sign"]
	if got == "" {
		return false
	}

	// 过滤空值与 sign 字段,按 key 升序
	keys := make([]string, 0, len(params))
	for k, v := range params {
		if v == "" || k == "sign" {
			continue
		}
		keys = append(keys, k)
	}
	sort.Strings(keys)

	var b strings.Builder
	for i, k := range keys {
		if i > 0 {
			b.WriteString("&")
		}
		b.WriteString(k)
		b.WriteString("=")
		b.WriteString(params[k])
	}

	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(b.String()))
	expected := hex.EncodeToString(mac.Sum(nil))
	return hmac.Equal([]byte(expected), []byte(got))
}

File: php/src/Namespaces/VCard.php
<?php
/**
 * 虚拟信用卡命名空间.
 *
 * 对应平台「虚拟信用卡」模块的 v1 公开接口.
 *
 * 鉴权分两档(凭据 = MPK 商户密钥 api_key / api_secret,与 Account 共用):
 *
 *   1) 只读 / 普通操作 —— X-API-Key + X-Secret-Key 双 header(同 Energy::authHeaders()).
 *      info / bins / cards / transactions / orders / remark
 *
 *   2) 敏感操作(读卡明文 / 开卡 / 充值 / 注销)—— HMAC-SHA256 头签名:
 *      payload = ts . nonce . METHOD . path . rawQuery . body
 *      sig     = hash_hmac('sha256', payload, api_secret)  (小写 hex)
 *      头:X-Key-ID / X-Timestamp / X-Nonce / X-Signature
 *      关键:POST body 先 json_encode 成字符串,对该串签名,再原样作为请求体发送
 *           (Http::request 支持传字符串 body 绕过二次 json_encode,保证签名字节一致).
 *
 * endpoint 一览:
 *
 *   GET  /api/v1/vcard/info                       getInfo               双密钥
 *   GET  /api/v1/vcard/bins                        listBins              双密钥
 *   GET  /api/v1/vcard/cards                       listCards             双密钥
 *   GET  /api/v1/vcard/cards/:id/transactions      getCardTransactions   双密钥
 *   GET  /api/v1/vcard/orders                       listOrders            双密钥
 *   GET  /api/v1/vcard/orders/:id                   getOrder              双密钥
 *   PUT  /api/v1/vcard/cards/:id/remark             updateCardRemark      双密钥
 *   GET  /api/v1/vcard/cards/:id/details            getCardDetails        HMAC 签名
 *   GET  /api/v1/vcard/cards/:id/code               getCardCode           HMAC 签名
 *   POST /api/v1/vcard/cards                         openCard              HMAC 签名
 *   POST /api/v1/vcard/cards/:id/recharge           rechargeCard          HMAC 签名
 *   POST /api/v1/vcard/cards/:id/cancel             cancelCard            HMAC 签名(无 body)
 *
 * 另提供静态方法 verifyWebhook() 校验平台 webhook 回调签名.
 */

declare(strict_types=1);

namespace NexCore\Namespaces;

use NexCore\Client;
use NexCore\NexCoreError;

class VCard
{
    private Client $client;

    public function __construct(Client $c) { $this->client = $c; }

    // ---------- 鉴权 ----------

    /**
     * 双密钥鉴权头(只读 / 普通操作用).
     *
     * @return list<string>
     * @throws NexCoreError 凭据未配置
     */
    private function authHeaders(): array
    {
        $k = $this->client->get('api_key');
        $s = $this->client->get('api_secret');
        if (!$k || !$s) {
            throw new NexCoreError('api_key / api_secret not configured', -1);
        }
        return ["X-API-Key: $k", "X-Secret-Key: $s"];
    }

    /**
     * HMAC 头签名统一发请求(敏感操作用).
     *
     * payload = ts . nonce . METHOD . path . rawQuery . body
     *   - ts:       (string)time()(秒级 unix 时间戳)
     *   - nonce:    bin2hex(random_bytes(8))(一次性随机串)
     *   - METHOD:   HTTP 方法大写
     *   - path:     请求路径(含 :id,不含 query)
     *   - rawQuery: 这些签名端点均无 query,固定 ""
     *   - body:     实际发送的 JSON 字符串(GET / 无 body 为 "")
     *
     * @param string $method
     * @param string $path
     * @param array<string, mixed>|null $body POST body 业务参数;GET / 无 body 传 null
     * @return array<mixed>
     * @throws NexCoreError
     */
    private function signedRequest(string $method, string $path, ?array $body = null): array
    {
        $apiKey    = $this->client->get('api_key');
        $apiSecret = $this->client->get('api_secret');
        if (!$apiKey || !$apiSecret) {
            throw new NexCoreError('api_key / api_secret not configured', -1);
        }

        $method   = strtoupper($method);
        $ts       = (string) time();
        $nonce    = bin2hex(random_bytes(8));
        $rawQuery = '';

        // body 先 json_encode 成字符串,签名与发送都用同一份,保证字节一致
        $bodyStr = '';
        if ($body !== null) {
            $bodyStr = json_encode($body, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
            if ($bodyStr === false) {
                throw new NexCoreError('vcard: json_encode body failed', -1);
            }
        }

        $payload = $ts . $nonce . $method . $path . $rawQuery . $bodyStr;
        $sig     = hash_hmac('sha256', $payload, $apiSecret);

        return $this->client->http->request($method, $path, [
            // 传字符串绕过 Http 再次 json_encode;无 body 时传 null 不发请求体
            'body'    => $body === null ? null : $bodyStr,
            'headers' => [
                'X-Key-ID: ' . $apiKey,
                'X-Timestamp: ' . $ts,
                'X-Nonce: ' . $nonce,
                'X-Signature: ' . $sig,
            ],
        ]);
    }

    // ---------- 双密钥(只读 / 普通操作) ----------

    /**
     * 虚拟卡平台信息(可开卡类型 / 费率 / 限额等).
     *
     * GET /api/v1/vcard/info
     *
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function getInfo(): array
    {
        return $this->client->http->request('GET', '/api/v1/vcard/info', [
            'headers' => $this->authHeaders(),
        ]);
    }

    /**
     * 列出可用卡 BIN(卡头).
     *
     * GET /api/v1/vcard/bins
     *
     * @return array<string, mixed> {list, ...}
     * @throws NexCoreError
     */
    public function listBins(): array
    {
        return $this->client->http->request('GET', '/api/v1/vcard/bins', [
            'headers' => $this->authHeaders(),
        ]);
    }

    /**
     * 列出名下虚拟卡.
     *
     * GET /api/v1/vcard/cards
     *
     * @return array<string, mixed> {list, total, ...}
     * @throws NexCoreError
     */
    public function listCards(): array
    {
        return $this->client->http->request('GET', '/api/v1/vcard/cards', [
            'headers' => $this->authHeaders(),
        ]);
    }

    /**
     * 查询单卡交易流水.
     *
     * GET /api/v1/vcard/cards/:id/transactions
     *
     * @param string $cardId 卡 ID
     * @return array<string, mixed> {list, ...}
     * @throws NexCoreError
     */
    public function getCardTransactions(string $cardId): array
    {
        if (!$cardId) {
            throw new NexCoreError('cardId is required', -1);
        }
        return $this->client->http->request('GET', "/api/v1/vcard/cards/$cardId/transactions", [
            'headers' => $this->authHeaders(),
        ]);
    }

    /**
     * 列出开卡 / 充值订单.
     *
     * GET /api/v1/vcard/orders
     *
     * @param array{
     *     page?: int,
     *     page_size?: int,
     *     status?: string,
     *     order_type?: string,
     * } $query
     * @return array<string, mixed> {list, total, page, ...}
     * @throws NexCoreError
     */
    public function listOrders(array $query = []): array
    {
        return $this->client->http->request('GET', '/api/v1/vcard/orders', [
            'headers' => $this->authHeaders(),
            'query'   => $query,
        ]);
    }

    /**
     * 查询单笔订单.
     *
     * GET /api/v1/vcard/orders/:id
     *
     * @param string $orderId 订单 ID
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function getOrder(string $orderId): array
    {
        if (!$orderId) {
            throw new NexCoreError('orderId is required', -1);
        }
        return $this->client->http->request('GET', "/api/v1/vcard/orders/$orderId", [
            'headers' => $this->authHeaders(),
        ]);
    }

    /**
     * 修改卡备注.
     *
     * PUT /api/v1/vcard/cards/:id/remark
     *
     * @param string $cardId 卡 ID
     * @param string $remark 新备注
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function updateCardRemark(string $cardId, string $remark): array
    {
        if (!$cardId) {
            throw new NexCoreError('cardId is required', -1);
        }
        return $this->client->http->request('PUT', "/api/v1/vcard/cards/$cardId/remark", [
            'headers' => $this->authHeaders(),
            'body'    => ['remark' => $remark],
        ]);
    }

    // ---------- HMAC 签名(敏感操作) ----------

    /**
     * 读取卡明文信息(卡号 / 有效期 / 持卡人等).
     *
     * GET /api/v1/vcard/cards/:id/details
     *
     * @param string $cardId 卡 ID
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function getCardDetails(string $cardId): array
    {
        if (!$cardId) {
            throw new NexCoreError('cardId is required', -1);
        }
        return $this->signedRequest('GET', "/api/v1/vcard/cards/$cardId/details");
    }

    /**
     * 读取卡安全码(CVV).
     *
     * GET /api/v1/vcard/cards/:id/code
     *
     * @param string $cardId 卡 ID
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function getCardCode(string $cardId): array
    {
        if (!$cardId) {
            throw new NexCoreError('cardId is required', -1);
        }
        return $this->signedRequest('GET', "/api/v1/vcard/cards/$cardId/code");
    }

    /**
     * 开卡.
     *
     * POST /api/v1/vcard/cards
     *
     * @param array{
     *     bin_platform_id: int,   卡 BIN 平台 ID(必填,listBins 返回)
     *     amount: float,          开卡金额(必填,> 0)
     * } $params
     * @return array<string, mixed> {order_id, status, total_cost}
     * @throws NexCoreError
     */
    public function openCard(array $params): array
    {
        return $this->signedRequest('POST', '/api/v1/vcard/cards', $params);
    }

    /**
     * 卡充值.
     *
     * POST /api/v1/vcard/cards/:id/recharge
     *
     * @param string $cardId 卡 ID
     * @param array<string, mixed> $params 充值参数 {amount, ...}
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function rechargeCard(string $cardId, array $params): array
    {
        if (!$cardId) {
            throw new NexCoreError('cardId is required', -1);
        }
        return $this->signedRequest('POST', "/api/v1/vcard/cards/$cardId/recharge", $params);
    }

    /**
     * 注销卡(无 body).
     *
     * POST /api/v1/vcard/cards/:id/cancel
     *
     * @param string $cardId 卡 ID
     * @return array<string, mixed>
     * @throws NexCoreError
     */
    public function cancelCard(string $cardId): array
    {
        if (!$cardId) {
            throw new NexCoreError('cardId is required', -1);
        }
        return $this->signedRequest('POST', "/api/v1/vcard/cards/$cardId/cancel");
    }

    // ---------- Webhook 验签 ----------

    /**
     * 校验平台 webhook 回调签名.
     *
     * 复刻后端签名算法:取所有非空、非 'sign' 字段,按 key 升序(ksort)拼成
     * "k1=v1&k2=v2",用商户 secret 做 HMAC-SHA256(小写 hex),与回调里的
     * sign 字段用 hash_equals 常量时间比较防时序攻击.
     *
     * 注:回调体一般含 sign_ts / nonce 字段,业务方应额外校验 sign_ts 时效
     *     (如 5 分钟内)并对 nonce 做去重,以防重放攻击.本方法只验签名正确性.
     *
     * @param array<string, mixed> $params 回调 JSON 完整解码后的 array(含 sign)
     * @param string $secret 商户密钥(api_secret)
     * @return bool true=签名正确可信;false=签名错误 / 缺失,应拒绝该回调
     */
    public static function verifyWebhook(array $params, string $secret): bool
    {
        $sign = $params['sign'] ?? null;
        if (!$sign || !is_string($sign)) {
            return false;
        }
        unset($params['sign']);
        $params = array_filter($params, fn($v) => $v !== '' && $v !== null);
        ksort($params);

        $pairs = [];
        foreach ($params as $k => $v) {
            $pairs[] = "$k=$v";
        }
        $expected = hash_hmac('sha256', implode('&', $pairs), $secret);
        return hash_equals($expected, $sign);
    }
}
